src/blockcipher/aes_ctr.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright The ocicrypt Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::blockcipher::{Finalizer, LayerBlockCipher, LayerBlockCipherOptions};
+use anyhow::{anyhow, Result};
+use ctr::cipher::generic_array::GenericArray;
+use ctr::cipher::{NewCipher, StreamCipher};
+use hmac::{Hmac, Mac, NewMac};
+use rand::{thread_rng, Rng};
+use sha2::Sha256;
+use std::io::Read;
+
+const AES256_KEY_SIZE: usize = 32;
+const AES256_NONCE_SIZE: usize = 16;
+type Aes256Ctr = ctr::Ctr128BE<aes::Aes256>;
+type HmacSha256 = Hmac<Sha256>;
+
+/// AESCTRLayerBlockCipher implements the AES CTR stream cipher
+#[derive(Clone)]
+pub struct AESCTRBlockCipher<R: Read> {
+    pub key_len: usize,
+    pub reader: Option<R>,
+    pub encrypt: bool,
+    pub cipher: Option<Aes256Ctr>,
+    pub hmac: Option<HmacSha256>,
+    pub exp_hmac: Vec<u8>,
+    pub done_encrypting: bool,
+}
+
+impl<R: Read> Read for AESCTRBlockCipher<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        if self.done_encrypting {
+            return Ok(0);
+        }
+
+        let read_len = self.reader.as_mut().unwrap().read(buf)?;
+        if read_len == 0 {
+            self.done_encrypting = true;
+        }
+
+        if !self.encrypt {
+            self.hmac.as_mut().unwrap().update(&buf[0..read_len]);
```

Comment:
Could we please avoid unwraping and do something like that instead:

```Rust
if !self.encrypt {
   if let Some(hmac) = self.hmac.as_mut() {
       hmac.update(&buf[0..read_len]);
   }
```

Reply:
Thanks, I'll fix it.